File: go/personality.go
// personality.go
//
// Ports CircleAI.Personality:
//   Persona / FormalityRange / PrivacyLevel      (Persona.cs)
//   IPersonaProvider                             (IPersonaProvider.cs)
//   JsonPersonaProvider                          (JsonPersonaProvider.cs)
//   IPersonaConflictResolver / DeclaredWinsResolver / LearnedWinsResolver (IPersonaConflictResolver.cs)
//   PersonaPromptBuilder.BuildSystemHint         (PersonaPromptBuilder.cs)
//
// PersonaState (the AI's LEARNED model) comes from memory.go — the resolvers
// bridge the declared Persona (here) against it. The prompt builder's
// prompt-injection defence (JSON-encode every user string) is preserved via
// encoding/json string marshalling, which escapes quotes/newlines/control
// characters exactly as the C# JsonSerializer.Serialize(value) does.

package circleai

import (
	"context"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"
)

// PrivacyLevel is the declared privacy posture. Ports the PrivacyLevel enum
// (stable ordinals: Strict=0, Balanced=1, Open=2).
type PrivacyLevel int

const (
	// PrivacyLevelStrict = minimum retention, no proactive surfacing.
	PrivacyLevelStrict PrivacyLevel = 0
	// PrivacyLevelBalanced is the default posture.
	PrivacyLevelBalanced PrivacyLevel = 1
	// PrivacyLevelOpen = maximum retention, share across surfaces.
	PrivacyLevelOpen PrivacyLevel = 2
)

// String renders the C# enum member name (used for JSON round-trip parity).
func (p PrivacyLevel) String() string {
	switch p {
	case PrivacyLevelStrict:
		return "Strict"
	case PrivacyLevelOpen:
		return "Open"
	default:
		return "Balanced"
	}
}

// privacyLevelFromString parses a PrivacyLevel enum member name (JSON form).
func privacyLevelFromString(s string) PrivacyLevel {
	switch s {
	case "Strict":
		return PrivacyLevelStrict
	case "Open":
		return PrivacyLevelOpen
	default:
		return PrivacyLevelBalanced
	}
}

// MarshalJSON emits the enum member name (JsonStringEnumConverter parity).
func (p PrivacyLevel) MarshalJSON() ([]byte, error) { return json.Marshal(p.String()) }

// UnmarshalJSON accepts the enum member name or an integer ordinal.
func (p *PrivacyLevel) UnmarshalJSON(b []byte) error {
	var s string
	if err := json.Unmarshal(b, &s); err == nil {
		*p = privacyLevelFromString(s)
		return nil
	}
	var n int
	if err := json.Unmarshal(b, &n); err != nil {
		return err
	}
	*p = PrivacyLevel(n)
	return nil
}

// FormalityRange declares bounds on conversational formality. Ports
// FormalityRange. Allowed values: "casual", "neutral", "formal".
type FormalityRange struct {
	Floor   string `json:"Floor"`
	Ceiling string `json:"Ceiling"`
}

// Persona is the user-declared persona artefact. Ports the Persona record.
type Persona struct {
	ID              uuid.UUID      `json:"Id"`
	DisplayName     string         `json:"DisplayName"`
	Pronouns        *string        `json:"Pronouns,omitempty"`
	IdentityTags    []string       `json:"IdentityTags"`
	Values          []string       `json:"Values"`
	Taboos          []string       `json:"Taboos"`
	PreferredLocale string         `json:"PreferredLocale"`
	VoicePreference *string        `json:"VoicePreference,omitempty"`
	Formality       FormalityRange `json:"Formality"`
	Privacy         PrivacyLevel   `json:"Privacy"`
	CreatedAt       time.Time      `json:"CreatedAt"`
	UpdatedAt       time.Time      `json:"UpdatedAt"`
}

// NewPersona creates a Persona with balanced defaults, an unconstrained
// formality range ("casual".."formal"), and now timestamps. Ports
// Persona.Create. Returns an error on empty displayName/locale.
func NewPersona(displayName, locale string) (Persona, error) {
	if strings.TrimSpace(displayName) == "" {
		return Persona{}, errors.New("displayName required")
	}
	if strings.TrimSpace(locale) == "" {
		return Persona{}, errors.New("locale required")
	}
	now := time.Now().UTC()
	return Persona{
		ID:              uuid.New(),
		DisplayName:     displayName,
		Pronouns:        nil,
		IdentityTags:    []string{},
		Values:          []string{},
		Taboos:          []string{},
		PreferredLocale: locale,
		VoicePreference: nil,
		Formality:       FormalityRange{Floor: "casual", Ceiling: "formal"},
		Privacy:         PrivacyLevelBalanced,
		CreatedAt:       now,
		UpdatedAt:       now,
	}, nil
}

// ---------------------------------------------------------------------------
// IPersonaConflictResolver (IPersonaConflictResolver.cs)
// ---------------------------------------------------------------------------

// IPersonaConflictResolver reconciles a declared Persona with the learned
// PersonaState. Ports IPersonaConflictResolver. Implementations are
// deterministic and never mutate their inputs.
type IPersonaConflictResolver interface {
	Resolve(declared Persona, learned PersonaState) Persona
}

// DeclaredWinsResolver clamps learned formality into the declared range; the
// declared record is otherwise authoritative. Ports DeclaredWinsResolver.
type DeclaredWinsResolver struct{}

// Resolve applies the declared-wins policy. Ports Resolve.
func (DeclaredWinsResolver) Resolve(declared Persona, learned PersonaState) Persona {
	clamped := clampFormality(learned.Formality, declared.Formality)
	if clamped == learned.Formality {
		return declared
	}
	var rng FormalityRange
	switch clamped {
	case "casual":
		rng = FormalityRange{Floor: "casual", Ceiling: declared.Formality.Ceiling}
	case "formal":
		rng = FormalityRange{Floor: declared.Formality.Floor, Ceiling: "formal"}
	default:
		rng = declared.Formality
	}
	out := declared
	out.Formality = rng
	return out
}

func clampFormality(learned string, rng FormalityRange) string {
	learnedRank := formalityRank(learned)
	floorRank := formalityRank(rng.Floor)
	ceilingRank := formalityRank(rng.Ceiling)
	if floorRank > ceilingRank {
		return rng.Floor
	}
	if learnedRank < floorRank {
		return rng.Floor
	}
	if learnedRank > ceilingRank {
		return rng.Ceiling
	}
	return learned
}

func formalityRank(formality string) int {
	switch formality {
	case "casual":
		return 0
	case "formal":
		return 2
	default:
		return 1 // neutral / unknown
	}
}

var _ IPersonaConflictResolver = DeclaredWinsResolver{}

// LearnedWinsResolver passes the declared persona through so identity, taboos,
// and values stay intact; the learned formality/locale/verbosity are applied
// separately by the prompt builder. Ports LearnedWinsResolver.
type LearnedWinsResolver struct{}

// Resolve returns the declared persona unchanged. Ports Resolve.
func (LearnedWinsResolver) Resolve(declared Persona, _ PersonaState) Persona {
	return declared
}

var _ IPersonaConflictResolver = LearnedWinsResolver{}

// ---------------------------------------------------------------------------
// IPersonaProvider (IPersonaProvider.cs)
// ---------------------------------------------------------------------------

// IPersonaProvider persists and retrieves user-declared Persona documents.
// Ports IPersonaProvider. ExportAllAsync (IAsyncEnumerable) is a slice-returner.
type IPersonaProvider interface {
	Get(ctx context.Context, userID string) (Persona, bool, error)
	Save(ctx context.Context, userID string, persona Persona) (Persona, error)
	Exists(ctx context.Context, userID string) (bool, error)
	ExportAll(ctx context.Context) ([]Persona, error)
}

// ---------------------------------------------------------------------------
// JsonPersonaProvider (JsonPersonaProvider.cs)
// ---------------------------------------------------------------------------

// JsonPersonaProvider stores each persona as {root}/{userId}.persona.json.
// Atomic write-then-rename; per-userId locking. Ports JsonPersonaProvider.
type JsonPersonaProvider struct {
	rootDirectory string
	locksMu       sync.Mutex
	locks         map[string]*sync.Mutex
}

// NewJsonPersonaProvider creates a provider rooted at rootDirectory, creating
// the directory if needed. Errors on empty root / mkdir failure.
func NewJsonPersonaProvider(rootDirectory string) (*JsonPersonaProvider, error) {
	if strings.TrimSpace(rootDirectory) == "" {
		return nil, errors.New("rootDirectory required")
	}
	if err := os.MkdirAll(rootDirectory, 0o755); err != nil {
		return nil, err
	}
	return &JsonPersonaProvider{rootDirectory: rootDirectory, locks: make(map[string]*sync.Mutex)}, nil
}

func (p *JsonPersonaProvider) lockFor(userID string) *sync.Mutex {
	p.locksMu.Lock()
	defer p.locksMu.Unlock()
	g, ok := p.locks[userID]
	if !ok {
		g = &sync.Mutex{}
		p.locks[userID] = g
	}
	return g
}

func (p *JsonPersonaProvider) personaPath(userID string) string {
	safe := sanitizeFileName(userID)
	if strings.TrimSpace(safe) == "" {
		safe = "default"
	}
	return filepath.Join(p.rootDirectory, safe+".persona.json")
}

// Get loads the persona for userID. Ports GetAsync.
func (p *JsonPersonaProvider) Get(ctx context.Context, userID string) (Persona, bool, error) {
	if strings.TrimSpace(userID) == "" {
		return Persona{}, false, errors.New("userId required")
	}
	path := p.personaPath(userID)
	if _, err := os.Stat(path); errors.Is(err, os.ErrNotExist) {
		return Persona{}, false, nil
	}
	gate := p.lockFor(userID)
	gate.Lock()
	defer gate.Unlock()
	data, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return Persona{}, false, nil
		}
		return Persona{}, false, err
	}
	var persona Persona
	if err := json.Unmarshal(data, &persona); err != nil {
		return Persona{}, false, err
	}
	return persona, true, nil
}

// Save persists persona (refreshing UpdatedAt) atomically. Ports SaveAsync.
func (p *JsonPersonaProvider) Save(ctx context.Context, userID string, persona Persona) (Persona, error) {
	if strings.TrimSpace(userID) == "" {
		return Persona{}, errors.New("userId required")
	}
	refreshed := persona
	refreshed.UpdatedAt = time.Now().UTC()
	target := p.personaPath(userID)
	tmp := target + "." + uuidNoDashes(uuid.New()) + ".tmp"

	data, err := json.MarshalIndent(refreshed, "", "  ")
	if err != nil {
		return Persona{}, err
	}

	gate := p.lockFor(userID)
	gate.Lock()
	defer gate.Unlock()
	if err := os.WriteFile(tmp, data, 0o644); err != nil {
		_ = os.Remove(tmp)
		return Persona{}, err
	}
	if err := os.Rename(tmp, target); err != nil {
		_ = os.Remove(tmp)
		return Persona{}, err
	}
	return refreshed, nil
}

// Exists reports whether a persona is stored for userID. Ports ExistsAsync.
func (p *JsonPersonaProvider) Exists(ctx context.Context, userID string) (bool, error) {
	if strings.TrimSpace(userID) == "" {
		return false, errors.New("userId required")
	}
	_, err := os.Stat(p.personaPath(userID))
	if errors.Is(err, os.ErrNotExist) {
		return false, nil
	}
	if err != nil {
		return false, err
	}
	return true, nil
}

// ExportAll streams every stored persona, skipping corrupt records. Ports
// ExportAllAsync.
func (p *JsonPersonaProvider) ExportAll(ctx context.Context) ([]Persona, error) {
	out := make([]Persona, 0)
	entries, err := os.ReadDir(p.rootDirectory)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return out, nil
		}
		return nil, err
	}
	for _, entry := range entries {
		if ctx.Err() != nil {
			return nil, ctx.Err()
		}
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".persona.json") {
			continue
		}
		data, rerr := os.ReadFile(filepath.Join(p.rootDirectory, entry.Name()))
		if rerr != nil {
			continue
		}
		var persona Persona
		if json.Unmarshal(data, &persona) != nil {
			continue
		}
		out = append(out, persona)
	}
	return out, nil
}

// sanitizeFileName replaces filesystem-invalid characters with '_' (mirrors the
// C# Path.GetInvalidFileNameChars join).
func sanitizeFileName(name string) string {
	invalid := "\x00<>:\"/\\|?*"
	return strings.Map(func(r rune) rune {
		if r < 32 || strings.ContainsRune(invalid, r) {
			return '_'
		}
		return r
	}, name)
}

var _ IPersonaProvider = (*JsonPersonaProvider)(nil)

// ---------------------------------------------------------------------------
// PersonaPromptBuilder (PersonaPromptBuilder.cs)
// ---------------------------------------------------------------------------

// BuildPersonaSystemHint renders persona into a compact system-prompt hint, or
// returns "" when the persona is effectively default. Every user string is
// JSON-quoted as a prompt-injection defence. Ports
// PersonaPromptBuilder.BuildSystemHint.
func BuildPersonaSystemHint(persona Persona) string {
	if isPersonaEffectivelyDefault(persona) {
		return ""
	}
	var sb strings.Builder
	sb.WriteString("[Persona]")
	sb.WriteString("\nYou are speaking with ")
	sb.WriteString(personaQuote(persona.DisplayName))
	sb.WriteByte('.')

	if persona.Pronouns != nil && strings.TrimSpace(*persona.Pronouns) != "" {
		sb.WriteString(" They identify as ")
		sb.WriteString(personaQuote(*persona.Pronouns))
		sb.WriteByte('.')
	}

	sb.WriteString("\nThey prefer responses in ")
	sb.WriteString(personaQuote(persona.PreferredLocale))
	sb.WriteString(", tone between ")
	sb.WriteString(personaQuote(persona.Formality.Floor))
	sb.WriteString(" and ")
	sb.WriteString(personaQuote(persona.Formality.Ceiling))
	sb.WriteByte('.')

	if len(persona.IdentityTags) > 0 {
		sb.WriteString("\nIdentity tags: ")
		sb.WriteString(personaQuoteList(persona.IdentityTags))
		sb.WriteByte('.')
	}
	if len(persona.Values) > 0 {
		sb.WriteString("\nTheir declared values: ")
		sb.WriteString(personaQuoteList(persona.Values))
		sb.WriteByte('.')
	}
	if len(persona.Taboos) > 0 {
		sb.WriteString("\nAvoid: ")
		sb.WriteString(personaQuoteList(persona.Taboos))
		sb.WriteByte('.')
	}
	if persona.VoicePreference != nil && strings.TrimSpace(*persona.VoicePreference) != "" {
		sb.WriteString("\nPreferred voice tag: ")
		sb.WriteString(personaQuote(*persona.VoicePreference))
		sb.WriteByte('.')
	}
	if persona.Privacy == PrivacyLevelStrict {
		sb.WriteString("\nPrivacy: strict — minimize stored signals, do not surface personal context proactively, and never share personal context across surfaces without explicit prompt.")
	} else if persona.Privacy == PrivacyLevelOpen {
		sb.WriteString("\nPrivacy: open — the user has authorised broader retention and proactive surfacing.")
	}
	return sb.String()
}

func isPersonaEffectivelyDefault(p Persona) bool {
	pronounsEmpty := p.Pronouns == nil || strings.TrimSpace(*p.Pronouns) == ""
	voiceEmpty := p.VoicePreference == nil || strings.TrimSpace(*p.VoicePreference) == ""
	return pronounsEmpty &&
		len(p.IdentityTags) == 0 &&
		len(p.Values) == 0 &&
		len(p.Taboos) == 0 &&
		voiceEmpty &&
		p.Privacy == PrivacyLevelBalanced &&
		p.Formality.Floor == "casual" &&
		p.Formality.Ceiling == "formal"
}

// personaQuote JSON-encodes value into a quoted literal (prompt-injection
// defence). Uses a non-HTML-escaping encoder to match UnsafeRelaxedJsonEscaping.
func personaQuote(value string) string {
	var sb strings.Builder
	enc := json.NewEncoder(&sb)
	enc.SetEscapeHTML(false)
	_ = enc.Encode(value)
	// Encode appends a trailing newline; trim it.
	return strings.TrimRight(sb.String(), "\n")
}

func personaQuoteList(items []string) string {
	if len(items) == 0 {
		return ""
	}
	parts := make([]string, len(items))
	for i, it := range items {
		parts[i] = personaQuote(it)
	}
	return strings.Join(parts, ", ")
}

File: c/src/personality.c
/*
 * personality.c — CircleAI.Personality (C11 port).
 *
 * The user-declared Persona artefact, an in-memory provider keyed by userId, the
 * two conflict resolvers (formality clamp / pass-through), and the prompt builder
 * that JSON-quotes every user string. Deterministic. Pure C11 + libc. Consumes
 * memory_brain.h's ca_uuid_v4 + memory.h's ca_persona_state_t. No pthreads.
 */

#include "circle_ai/personality.h"
#include "circle_ai/security.h" /* ca_uuid_v4, CA_UUID_STR_LEN */
#include "board_common.h"

/* ── record helpers ─────────────────────────────────────────────────────── */

void ca_persona_free(ca_persona_t *p) {
    if (!p) return;
    free(p->id);
    free(p->display_name);
    free(p->pronouns);
    cab_strv_free(p->identity_tags, p->identity_tag_count);
    cab_strv_free(p->values, p->value_count);
    cab_strv_free(p->taboos, p->taboo_count);
    free(p->preferred_locale);
    free(p->voice_preference);
    free(p->formality.floor);
    free(p->formality.ceiling);
    memset(p, 0, sizeof(*p));
}
void ca_persona_free_array(ca_persona_t *arr, size_t count) {
    if (!arr) return;
    for (size_t i = 0; i < count; ++i) ca_persona_free(&arr[i]);
    free(arr);
}

bool ca_persona_copy(ca_persona_t *dst, const ca_persona_t *src) {
    memset(dst, 0, sizeof(*dst));
    dst->privacy       = src->privacy;
    dst->created_at_ms = src->created_at_ms;
    dst->updated_at_ms = src->updated_at_ms;
    dst->id               = cab_strdup_empty(src->id);
    dst->display_name     = cab_strdup_empty(src->display_name);
    dst->preferred_locale = cab_strdup_empty(src->preferred_locale);
    dst->pronouns         = src->pronouns ? cab_strdup(src->pronouns) : NULL;
    dst->voice_preference = src->voice_preference ? cab_strdup(src->voice_preference) : NULL;
    dst->formality.floor   = cab_strdup_empty(src->formality.floor);
    dst->formality.ceiling = cab_strdup_empty(src->formality.ceiling);
    if (!dst->id || !dst->display_name || !dst->preferred_locale ||
        !dst->formality.floor || !dst->formality.ceiling ||
        (src->pronouns && !dst->pronouns) ||
        (src->voice_preference && !dst->voice_preference)) {
        ca_persona_free(dst); return false;
    }
    if (!cab_strv_copy(&dst->identity_tags, src->identity_tags, src->identity_tag_count) ||
        !cab_strv_copy(&dst->values, src->values, src->value_count) ||
        !cab_strv_copy(&dst->taboos, src->taboos, src->taboo_count)) {
        ca_persona_free(dst); return false;
    }
    dst->identity_tag_count = src->identity_tag_count;
    dst->value_count = src->value_count;
    dst->taboo_count = src->taboo_count;
    return true;
}

bool ca_persona_create(const char *display_name, const char *locale,
                       int64_t now_ms, ca_persona_t *out) {
    if (out) memset(out, 0, sizeof(*out));
    if (!out || cab_is_ws(display_name) || cab_is_ws(locale)) return false;
    char uuid[CA_UUID_STR_LEN];
    ca_uuid_v4(uuid);
    out->id = cab_strdup_empty(uuid);
    out->display_name = cab_strdup_empty(display_name);
    out->preferred_locale = cab_strdup_empty(locale);
    out->pronouns = NULL;
    out->voice_preference = NULL;
    out->identity_tags = NULL; out->identity_tag_count = 0;
    out->values = NULL; out->value_count = 0;
    out->taboos = NULL; out->taboo_count = 0;
    out->formality.floor = cab_strdup_empty("casual");
    out->formality.ceiling = cab_strdup_empty("formal");
    out->privacy = CA_PRIVACY_BALANCED;
    out->created_at_ms = now_ms;
    out->updated_at_ms = now_ms;
    if (!out->id || !out->display_name || !out->preferred_locale ||
        !out->formality.floor || !out->formality.ceiling) {
        ca_persona_free(out); return false;
    }
    return true;
}

/* ── InMemoryPersonaProvider ────────────────────────────────────────────── */

typedef struct { char *user_id; ca_persona_t persona; } persona_entry_t;

struct ca_persona_provider {
    persona_entry_t *items; size_t count, cap;
};

ca_persona_provider_t *ca_persona_provider_create(void) {
    return (ca_persona_provider_t *)calloc(1, sizeof(ca_persona_provider_t));
}
void ca_persona_provider_destroy(ca_persona_provider_t *p) {
    if (!p) return;
    for (size_t i = 0; i < p->count; ++i) {
        free(p->items[i].user_id);
        ca_persona_free(&p->items[i].persona);
    }
    free(p->items);
    free(p);
}

bool ca_persona_provider_get(const ca_persona_provider_t *p, const char *user_id,
                             ca_persona_t *out) {
    if (out) memset(out, 0, sizeof(*out));
    if (!p || cab_is_ws(user_id) || !out) return false;
    for (size_t i = 0; i < p->count; ++i)
        if (cab_ord_eq(p->items[i].user_id, user_id))
            return ca_persona_copy(out, &p->items[i].persona);
    return false;
}

int ca_persona_provider_save(ca_persona_provider_t *p, const char *user_id,
                             const ca_persona_t *persona, int64_t now_ms,
                             ca_persona_t *out) {
    if (out) memset(out, 0, sizeof(*out));
    if (!p || cab_is_ws(user_id) || !persona) return -1;
    ca_persona_t refreshed;
    if (!ca_persona_copy(&refreshed, persona)) return -1;
    refreshed.updated_at_ms = now_ms;

    for (size_t i = 0; i < p->count; ++i) {
        if (cab_ord_eq(p->items[i].user_id, user_id)) {
            ca_persona_free(&p->items[i].persona);
            /* store a copy, keep `refreshed` for the return */
            if (!ca_persona_copy(&p->items[i].persona, &refreshed)) {
                ca_persona_free(&refreshed); return -1;
            }
            if (out) { if (!ca_persona_copy(out, &refreshed)) { ca_persona_free(&refreshed); return -1; } }
            ca_persona_free(&refreshed);
            return 0;
        }
    }
    if (p->count == p->cap) {
        size_t nc = p->cap ? p->cap * 2 : 4;
        void *n = realloc(p->items, nc * sizeof(persona_entry_t));
        if (!n) { ca_persona_free(&refreshed); return -1; }
        p->items = (persona_entry_t *)n; p->cap = nc;
    }
    char *uid = cab_strdup_empty(user_id);
    if (!uid) { ca_persona_free(&refreshed); return -1; }
    p->items[p->count].user_id = uid;
    if (!ca_persona_copy(&p->items[p->count].persona, &refreshed)) {
        free(uid); ca_persona_free(&refreshed); return -1;
    }
    p->count++;
    if (out) { if (!ca_persona_copy(out, &refreshed)) { ca_persona_free(&refreshed); return -1; } }
    ca_persona_free(&refreshed);
    return 0;
}

bool ca_persona_provider_exists(const ca_persona_provider_t *p, const char *user_id) {
    if (!p || cab_is_ws(user_id)) return false;
    for (size_t i = 0; i < p->count; ++i)
        if (cab_ord_eq(p->items[i].user_id, user_id)) return true;
    return false;
}

ca_persona_t *ca_persona_provider_export_all(const ca_persona_provider_t *p,
                                             size_t *out_count) {
    if (!out_count) return NULL;
    if (!p) { *out_count = (size_t)-1; return NULL; }
    if (p->count == 0) { *out_count = 0; return NULL; }
    ca_persona_t *out = (ca_persona_t *)calloc(p->count, sizeof(*out));
    if (!out) { *out_count = (size_t)-1; return NULL; }
    for (size_t i = 0; i < p->count; ++i) {
        if (!ca_persona_copy(&out[i], &p->items[i].persona)) {
            ca_persona_free_array(out, i);
            *out_count = (size_t)-1; return NULL;
        }
    }
    *out_count = p->count;
    return out;
}

/* ── conflict resolvers ─────────────────────────────────────────────────── */

static int formality_rank(const char *f) {
    if (cab_ord_eq(f, "casual")) return 0;
    if (cab_ord_eq(f, "neutral")) return 1;
    if (cab_ord_eq(f, "formal")) return 2;
    return 1; /* unknown -> neutral */
}
static const char *formality_state_str(ca_formality_t f) {
    switch (f) {
        case CA_FORMALITY_CASUAL: return "casual";
        case CA_FORMALITY_FORMAL: return "formal";
        default: return "neutral";
    }
}
/* Clamp `learned` into [floor, ceiling]; if inverted, return floor. */
static const char *clamp_formality(const char *learned,
                                   const char *floor, const char *ceiling) {
    int lr = formality_rank(learned);
    int fr = formality_rank(floor);
    int cr = formality_rank(ceiling);
    if (fr > cr) return floor;
    if (lr < fr) return floor;
    if (lr > cr) return ceiling;
    return learned;
}

bool ca_persona_resolve_declared_wins(const ca_persona_t *declared,
                                      const ca_persona_state_t *learned,
                                      ca_persona_t *out) {
    if (out) memset(out, 0, sizeof(*out));
    if (!declared || !learned || !out) return false;
    if (!ca_persona_copy(out, declared)) return false;

    const char *lstr = formality_state_str(learned->formality);
    const char *clamped = clamp_formality(lstr, declared->formality.floor,
                                          declared->formality.ceiling);
    if (cab_ord_eq(clamped, lstr)) {
        /* learned was within bounds — declared unchanged */
        return true;
    }
    /* surface the clamped value by adjusting floor/ceiling */
    if (cab_ord_eq(clamped, "casual")) {
        char *nf = cab_strdup_empty("casual");
        if (!nf) { ca_persona_free(out); return false; }
        free(out->formality.floor);
        out->formality.floor = nf;
        /* ceiling stays declared.Ceiling (already copied) */
    } else if (cab_ord_eq(clamped, "formal")) {
        char *nc = cab_strdup_empty("formal");
        if (!nc) { ca_persona_free(out); return false; }
        free(out->formality.ceiling);
        out->formality.ceiling = nc;
    }
    return true;
}

bool ca_persona_resolve_learned_wins(const ca_persona_t *declared,
                                     const ca_persona_state_t *learned,
                                     ca_persona_t *out) {
    if (out) memset(out, 0, sizeof(*out));
    if (!declared || !learned || !out) return false;
    /* passes the declared persona through unchanged */
    return ca_persona_copy(out, declared);
}

/* ── PersonaPromptBuilder ───────────────────────────────────────────────── */

/* True when the persona carries nothing beyond Create() defaults. */
static bool persona_is_default(const ca_persona_t *p) {
    return cab_is_ws(p->pronouns) &&
           p->identity_tag_count == 0 &&
           p->value_count == 0 &&
           p->taboo_count == 0 &&
           cab_is_ws(p->voice_preference) &&
           p->privacy == CA_PRIVACY_BALANCED &&
           cab_ord_eq(p->formality.floor, "casual") &&
           cab_ord_eq(p->formality.ceiling, "formal");
}

/* Growable char buffer. */
typedef struct { char *buf; size_t len, cap; } sb_t;
static bool sb_reserve(sb_t *sb, size_t extra) {
    if (sb->len + extra + 1 > sb->cap) {
        size_t nc = sb->cap ? sb->cap : 64;
        while (nc < sb->len + extra + 1) nc *= 2;
        char *nb = (char *)realloc(sb->buf, nc);
        if (!nb) return false;
        sb->buf = nb; sb->cap = nc;
    }
    return true;
}
static bool sb_puts(sb_t *sb, const char *s) {
    size_t n = strlen(s);
    if (!sb_reserve(sb, n)) return false;
    memcpy(sb->buf + sb->len, s, n);
    sb->len += n;
    sb->buf[sb->len] = '\0';
    return true;
}
/* Append `raw` as a JSON string literal (relaxed escaping: quote + backslash +
 * control chars). */
static bool sb_put_quoted(sb_t *sb, const char *raw) {
    if (!raw) raw = "";
    if (!sb_reserve(sb, strlen(raw) * 6 + 2)) return false;
    char *p = sb->buf + sb->len;
    *p++ = '"';
    for (const unsigned char *c = (const unsigned char *)raw; *c; ++c) {
        switch (*c) {
            case '"':  *p++ = '\\'; *p++ = '"';  break;
            case '\\': *p++ = '\\'; *p++ = '\\'; break;
            case '\b': *p++ = '\\'; *p++ = 'b';  break;
            case '\f': *p++ = '\\'; *p++ = 'f';  break;
            case '\n': *p++ = '\\'; *p++ = 'n';  break;
            case '\r': *p++ = '\\'; *p++ = 'r';  break;
            case '\t': *p++ = '\\'; *p++ = 't';  break;
            default:
                if (*c < 0x20) {
                    static const char hex[] = "0123456789abcdef";
                    *p++ = '\\'; *p++ = 'u'; *p++ = '0'; *p++ = '0';
                    *p++ = hex[(*c >> 4) & 0xF]; *p++ = hex[*c & 0xF];
                } else {
                    *p++ = (char)*c;
                }
        }
    }
    *p++ = '"';
    *p = '\0';
    sb->len = (size_t)(p - sb->buf);
    return true;
}
static bool sb_put_quoted_list(sb_t *sb, char *const *items, size_t n) {
    for (size_t i = 0; i < n; ++i) {
        if (i > 0 && !sb_puts(sb, ", ")) return false;
        if (!sb_put_quoted(sb, items[i])) return false;
    }
    return true;
}

char *ca_persona_build_system_hint(const ca_persona_t *persona) {
    if (!persona) return NULL;
    if (persona_is_default(persona)) return cab_strdup_empty("");

    sb_t sb = {0};
    bool ok = true;
    ok = ok && sb_puts(&sb, "[Persona]");
    ok = ok && sb_puts(&sb, "\nYou are speaking with ");
    ok = ok && sb_put_quoted(&sb, persona->display_name);
    ok = ok && sb_puts(&sb, ".");

    if (ok && !cab_is_ws(persona->pronouns)) {
        ok = sb_puts(&sb, " They identify as ") &&
             sb_put_quoted(&sb, persona->pronouns) &&
             sb_puts(&sb, ".");
    }

    ok = ok && sb_puts(&sb, "\nThey prefer responses in ");
    ok = ok && sb_put_quoted(&sb, persona->preferred_locale);
    ok = ok && sb_puts(&sb, ", tone between ");
    ok = ok && sb_put_quoted(&sb, persona->formality.floor);
    ok = ok && sb_puts(&sb, " and ");
    ok = ok && sb_put_quoted(&sb, persona->formality.ceiling);
    ok = ok && sb_puts(&sb, ".");

    if (ok && persona->identity_tag_count > 0) {
        ok = sb_puts(&sb, "\nIdentity tags: ") &&
             sb_put_quoted_list(&sb, persona->identity_tags, persona->identity_tag_count) &&
             sb_puts(&sb, ".");
    }
    if (ok && persona->value_count > 0) {
        ok = sb_puts(&sb, "\nTheir declared values: ") &&
             sb_put_quoted_list(&sb, persona->values, persona->value_count) &&
             sb_puts(&sb, ".");
    }
    if (ok && persona->taboo_count > 0) {
        ok = sb_puts(&sb, "\nAvoid: ") &&
             sb_put_quoted_list(&sb, persona->taboos, persona->taboo_count) &&
             sb_puts(&sb, ".");
    }
    if (ok && !cab_is_ws(persona->voice_preference)) {
        ok = sb_puts(&sb, "\nPreferred voice tag: ") &&
             sb_put_quoted(&sb, persona->voice_preference) &&
             sb_puts(&sb, ".");
    }
    if (ok && persona->privacy == CA_PRIVACY_STRICT) {
        ok = sb_puts(&sb, "\nPrivacy: strict — minimize stored signals, do not "
                          "surface personal context proactively, and never share "
                          "personal context across surfaces without explicit prompt.");
    } else if (ok && persona->privacy == CA_PRIVACY_OPEN) {
        ok = sb_puts(&sb, "\nPrivacy: open — the user has authorised broader "
                          "retention and proactive surfacing.");
    }

    if (!ok) { free(sb.buf); return NULL; }
    if (!sb.buf) return cab_strdup_empty("");
    return sb.buf;
}
